crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -320,7 +349,74 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
                 ast::Singleton::True => Type::BooleanLiteral(true),
                 ast::Singleton::False => Type::BooleanLiteral(false),
             };
-            self.constraints.insert(symbol, ty);
+            let mut constraints = NarrowingConstraints::default();
+            constraints.insert(symbol, ty);
+            Some(constraints)
+        } else {
+            None
+        }
+    }
+
+    fn add_bool_op(
+        &mut self,
+        expr_bool_op: &ExprBoolOp,
+        expression: Expression<'db>,
+        is_positive: bool,
+    ) -> Option<NarrowingConstraints<'db>> {
+        match (expr_bool_op.op, is_positive) {
+            (BoolOp::And, true) | (BoolOp::Or, false) => {
+                let mut aggregated_constraints: Option<NarrowingConstraints<'db>> = None;
+                for sub_constraint in expr_bool_op
+                    .values
+                    .iter()
+                    .filter_map(|sub_expr| {
+                        self.evaluate_expression_node_constraint(sub_expr, expression, is_positive)
+                    })
+                    .collect::<Vec<_>>()
+                {
+                    if let Some(ref mut some_aggregated_constraints) = aggregated_constraints {
+                        merge_constraints(some_aggregated_constraints, sub_constraint, self.db);
+                    } else {
+                        aggregated_constraints = Some(sub_constraint);
+                    }
+                }
+                aggregated_constraints
+            }
+            (BoolOp::Or, true) | (BoolOp::And, false) => {
+                let sub_constraints = expr_bool_op
+                    .values
+                    .iter()
+                    .filter_map(|sub_expr| {
+                        self.evaluate_expression_node_constraint(sub_expr, expression, is_positive)
+                    })
+                    // In order to narrow down based on OR operator, all arms must create a constraint,
+                    // and all constraints must be of exactly one symbol.
+                    .filter(|x| x.len() == 1)
```

Comment:
I think this is too strict, and will mean we can't understand a more complex constraint like `if (isinstance(x, A) and isinstance(y, B)) or (isinstance(x, B) and isinstance(y, A)):`

In this case of course we have no way to represent the precise relationship between the types of `x` and `y`, but it is correct to narrow both `x` and `y` to `A | B`. Your current code will refuse to do that.

I think the code might be clearer here if you structure it more similarly to the above branch, rename the current `merge_constraints` function to `merge_constraints_and`, and write a new `merge_constraints_or` to use in this branch.

The `merge_constraints_or` function should union the constraints (instead of intersecting them like `merge_constraints_and` does), and should eliminate any variables not constrained by either side, or where the union results in `object` (these two cases are really the same: no constraint for a variable is implicitly "constrained to `object`").